rs-matter/src/transport/network/btp/context.rs

```diff
@@ -0,0 +1,378 @@
+            handshake_notif: Notification::new(),
+            available_notif: Notification::new(),
+            recv_notif: Notification::new(),
+            ack_notif: Notification::new(),
+            send_notif: Notification::new(),
+        }
+    }
+}
+
+impl<M> BtpContext<M>
+where
+    M: RawMutex,
+{
+    /// The `Btp` instance passes a closure of this method to the `GattPeripheral` implementation which is in use
+    /// so that the peripheral can report to it subscribe, unsubscribe and write events.
+    pub(crate) fn on_event(&self, event: GattPeripheralEvent) {
+        let result = match event {
+            GattPeripheralEvent::NotifySubscribed(address) => self.on_subscribe(address),
+            GattPeripheralEvent::NotifyUnsubscribed(address) => self.on_unsubscribe(address),
+            GattPeripheralEvent::Write {
+                address,
+                data,
+                gatt_mtu,
+            } => self.on_write(address, data, gatt_mtu),
+        };
+
+        if let Err(e) = result {
+            error!("Unexpected error in GATT callback: {e:?}");
+        }
+    }
+
+    /// Handles a write event to characteristic `C1` from the GATT peripheral.
+    fn on_write(&self, address: BtAddr, data: &[u8], gatt_mtu: Option<u16>) -> Result<(), Error> {
+        trace!("Received {data:02x?} bytes from {address}");
+
+        self.sessions.lock(|sessions| {
+            let mut sessions = sessions.borrow_mut();
+
+            if Session::is_handshake(data)? {
+                if sessions.len() >= MAX_BTP_SESSIONS {
+                    warn!("Too many BTP sessions, dropping a handshake request from address {address}");
+                } else {
+                    // Unwrap is safe because we checked the length above
+                    sessions
+                        .push(Session::process_rx_handshake(address, data, gatt_mtu)?)
+                        .unwrap();
+                }
+
+                Ok(())
+            } else {
+                let Some(index) = sessions
+                    .iter_mut()
+                    .position(|session| session.address() == address)
+                else {
+                    warn!("Dropping data from address {address} because there is no session for it");
+                    return Ok(());
+                };
+
+                let session = &mut sessions[index];
+                let result = session.process_rx_data(data);
+
+                if result.is_err() {
+                    sessions.swap_remove(index);
+                    error!("Dropping session {address} because of an error: {result:?}");
+                }
+
+                self.available_notif.notify();
+                self.recv_notif.notify();
+                self.ack_notif.notify();
+                self.send_notif.notify();
+
+                result
+            }
+        })
+    }
+
+    /// Handles a subscribe event to characteristic `C2` from the GATT peripheral.
```

Comment:
> since names are generic, how do we know on_write is only C1 and on_subscribe is only C2? would it make sense to pass in a characteristic identifier to these or name them differently?

* Passing a characteristic identifier does not make any sense, because handling writes on `C2` (and subscriptions on `C1`) is *not defined by the spec* and is therefore a meaningless operation. The implementor of `GattPeripheral` is (a) not even supposed to allow that (by putting proper permissions on the C1 and C2 characteristics) and (b) if that happens (for whatever reason) it is NOT supposed to reach the BTP code anyway (i.e. the GATT peripheral should error out to the peer)
* Renaming `on_subscribe` to `on_c2_subscribe` and `on_write` to `on_c1_write`
  * I'm _firmly against it_, as that would imply that:
    * `on_c1_subscribe` might happen and is a valid operation, but it is not
    * `on_c2_write` might happen and is a valid operation, but it is not

The peer subscribes to our indications (= our "writes") via the C2 characteristic, and sends us data ("writes" to us) via the C1 characteristic, and that's it. Everything else is undefined/error/deviating from the BTP protocol.

Reply:
Ok, closing. Reserving the right to say "I told you so" the first time we get bugs due to unexpected code paths in what looks like generic code :)